### PR TITLE
Add support for providing command information via php8 Attributes.

### PR DESCRIPTION
**Pull request by beejeebus**
### Disposition
This pull request:

- Adds a feature - [Adds support providing command information via php8 Attributes](https://github.com/consolidation/annotated-command/issues/215).
- Has tests that cover changes

### Description
Any additional information.

### CODE REVIEW

File: tests/src/ExampleAttributesCommandFile.php
Comment (greg-1-anderson) [inline tests/src/ExampleAttributesCommandFile.php:34]: This CommandAttributes structure looks pretty good to me. The only question I have is what do we want to name it? `#[CLI(...)]`?  CommandAttributes is also fine.

**Comment by greg-1-anderson** [comment]
Looks good. The other thing that I suddenly started wondering about is, would it make sense to suggest this directly in `symfony/symfony` for the Console component? Right now the Annotated Command library provides not just the functionality of its namesake, but also provides the command hook implementation. Would it be possible to implement hooks if command annotations were native in Symfony Console commands?

We could also consider implementing hooks with annotations.

We don't need to make these blockers for moving forward here (Symfony first would likely be too slow), but it would be good to have some aspirational ideas.

**Comment by beejeebus** [comment]
> The other thing that I suddenly started wondering about is, would it make sense to suggest this directly in symfony/symfony for the Console component?

yeah, that probably makes sense after this is merged.

**Comment by beejeebus** [comment]
> We could also consider implementing hooks with annotations.

happy to work on that as a follow up to this PR.

File: tests/src/ExampleAttributesCommandFile.php
Comment (beejeebus) [reply tests/src/ExampleAttributesCommandFile.php:34]: no strong feelings, so when this is ready to merge, let's bikeshed 🙃

**Comment by greg-1-anderson** [comment]
Could you change the [branch alias](https://github.com/consolidation/annotated-command/blob/main/composer.json#L57) to 5.x?  I'll fork a 4.x branch off of main before this is merged.

While we might eventually be able to backport this to the 4.x branch & maintain b/c with that version series (probably will?), I want to emphasize that the first implementation is not stable, and might change.

**Comment by beejeebus** [comment]
> Could you change the branch alias to 5.x? I'll fork a 4.x branch off of main before this is merged.

done.

**Comment by beejeebus** [comment]
moved this to ready for review. I realized that much of the original code could be thrown away 😄 

remaining todo:

* settle on naming for the Attributes class
* probably some more tests, not sure all of the required fields are supported.

**Comment by beejeebus** [comment]
Added support for params, options and annotations, so I think we're close to parity with annotation support.

@greg-1-anderson - what else would you like to see before this can be merged? Now is a good time to discuss names :-)

File: src/Parser/Internal/AttributesDocBlockParser.php
Comment (beejeebus) [inline src/Parser/Internal/AttributesDocBlockParser.php:83]: These two methods are cut and paste from `BespokeDocBlockParser`. Happy to make a base class or trait these can share, but didn't want to refactor too much in this PR.

File: tests/src/ExampleAttributesCommandFile.php
Comment (andypost) [reply tests/src/ExampleAttributesCommandFile.php:40]: As I recall multi-line attributes are not BC for syntax on PHP <8.0

Comment (andypost) [reply tests/src/ExampleAttributesCommandFile.php:40]: About naming DrushCommands looks better name

File: src/Parser/Internal/AttributesDocBlockParser.php
Comment (greg-1-anderson) [reply src/Parser/Internal/AttributesDocBlockParser.php:83]: We could reduce code duplication in a follow-on PR.

Comment (greg-1-anderson) [reply src/Parser/Internal/AttributesDocBlockParser.php:83]: Maybe just add a comment just in case.

File: tests/src/ExampleAttributesCommandFile.php
Comment (greg-1-anderson) [reply tests/src/ExampleAttributesCommandFile.php:40]: Regarding the color of the CommandAttributes shed, I think that the term "Command" is very general and could refer to a wide variety of different sorts of commands. I think I'd therefore prefer "CommandLine" or "CommandLineAttributes". Not sure I am still enamored of my initial suggestion "CLI".

Regarding multi-line attributes, I think that we should recommend them by default. I don't think anyone is going to want to use command line attributes at all unless they are being used in a context where php 8 is required. Otherwise, you'd need both the php 8 attribute syntax and the legacy syntax at the same time, and we'd have to make sure that annotated command would not get confused and register the same command twice in such a scenario. Better to use only the legacy syntax until php 8 is your minimum supported version.

Regarding the name "DrushCommands", whatever we choose here is how Robo commands and any CLI tool derived from Robo will be annotated, so we need a general name. As follow-on work we could consider allowing derived applications to extend our command line attribute class. I don't think it would be worth it to do that unless the app wanted to extend the definition, but in that instance I expect it would be preferable to just define a second annotation. I don't have a strong opinion on this, but it doesn't seem too important at the moment.

**Comment by greg-1-anderson** [comment]
It would be good to hear if @weitzman or @stovak had any comments on this.

**Comment by greg-1-anderson** [comment]
A created a [4.x](https://github.com/consolidation/annotated-command/tree/4.x) branch.

File: tests/src/ExampleAttributesCommandFile.php
Comment (beejeebus) [reply tests/src/ExampleAttributesCommandFile.php:40]: I went with `CommandLineAttributes`. renaming is a small change, so further discussion is welcome.

File: tests/src/ExampleAttributesCommandFile.php
Comment (weitzman) [inline tests/src/ExampleAttributesCommandFile.php:62]: what does annotations mean?

**Review by weitzman — COMMENTED** [review]
LGTM. Thanks much.

**Comment by andypost** [comment]
FYI there's compatibility layer for PHP 7.2+ https://github.com/spiral/attributes

File: tests/src/ExampleAttributesCommandFile.php
Comment (beejeebus) [reply tests/src/ExampleAttributesCommandFile.php:62]: this is the bucket for 'custom keys not explicitly supported'. perhaps i should rename to 'custom' or 'customAnnotations'?

Comment (beejeebus) [reply tests/src/ExampleAttributesCommandFile.php:62]: annotations is for 'custom fields we don't have an explicit handler for'.

should i rename this to 'customAnnotations' or similar?

Comment (weitzman) [reply tests/src/ExampleAttributesCommandFile.php:62]: I would go with `custom` or `extra`

**Comment by beejeebus** [comment]
Looks like I need someone else to kick the workflows...

**Comment by greg-1-anderson** [comment]
Needs an update to composer.lock.

The other question is, what should the minimum PHP version be for this branch? For sure we do not need to keep testing PHP 7.1 here. I think we should probably support 7.3+, e.g. so there can be a version of Drush that supports php annotations for Drush extensions that require PHP 8+, while also supporting the existing command syntax.

**Comment by greg-1-anderson** [comment]
Ah, nevermind, Drush 10 still supports PHP 7.1, so we should probably keep supporting that here too in case we want to use this in Drush 10 at some point.

If we are sure this will be Drush 11 only, then we can drop some PHP versions.

**Comment by weitzman** [comment]
Drush 11 only makes sense to me.

>

**Comment by beejeebus** [comment]
Update composer.lock.

**Comment by greg-1-anderson** [comment]
Looks like coding guidelines will need to run with php8 now

**Comment by greg-1-anderson** [comment]
This branch is supposed to break b/c. Comment out the b/c test, and update it to run against 5.0.0.  We can re-enable it when 5.0.0 exists.

**Comment by greg-1-anderson** [comment]
I guess now we need to decide if this branch is php 8 only. Could probably support older versions with judicious use of MarkTestSkipped. Would need to make sure that none of the example files with multi-line attributes are loaded on older versions of PHP.

That might be too much effort. I think I'm okay with PHP 8+ only for 5.x.  Anyone have any reason why that might be a problem?

**Comment by beejeebus** [comment]
I don't have any problem with PHP 8+ for 5.x, but I'd like to hear what @weitzman thinks...

**Comment by weitzman** [comment]
I'm ok with php8-only but it strikes me as a bit premature. MarkTestSkipped is a very fast copy/paste where needed.

**Comment by greg-1-anderson** [comment]
Is Drush 11 going to support PHP 7?  Drupal 10 will be PHP 8+

**Comment by greg-1-anderson** [comment]
Also, I don't think that MarkTestSkipped is adequate for PHP 7 support. If there's a multi-line attribute in a file, then PHP 7 will crash if the file is loaded. We'd have to make sure that the example files that include annotations are not included or autoloaded unless the php version is 8+

**Comment by weitzman** [comment]
Right - during tests. Thats an environment that we control. It sounds like it would be easy but I admit I have not looked into it. 

Drush 11 will probably be drupal 10+ so we are good there. I'll be happy with whatever you choose here.

**Comment by greg-1-anderson** [comment]
I certainly don't have any objection to supporting PHP 7 if @beejeebus fixes the tests; I just don't know that there's a need.

**Comment by beejeebus** [comment]
I can poke at making it work for PHP 7. If it turns out hard, I'll come back here and complain 😉

**Comment by stovak** [comment]
I've been struggling with this same question for Terminus. The first time you add an attribute, php7 can't load the file. It's going to need to either be php 7 or php 8 and 'nary the twain shall meet.

**Comment by beejeebus** [comment]
skipping attribute tests when PHP version is less than 8.0.0. seems to work.

**Comment by beejeebus** [comment]
ping @greg-1-anderson - can you kick the workflow? thanks!

**Comment by greg-1-anderson** [comment]
I invited you to be an r/o collaborator, so maybe it will stop asking for permission to run the tests if you accept that.

**Comment by greg-1-anderson** [comment]
I thought I made a comment about the `CI / Backward Compatibility (pull_request)` in the past, but I don't see it now.

This PR creates a new major version, so we expect bc to break. Comment out this test, change it to compare with a future 5.0.0 tag, and add a note to bring the check back once the stable tag is created.

File: composer.json
Comment (greg-1-anderson) [inline composer.json:47]: 👍

**Review by greg-1-anderson — APPROVED** [review]
LGTM, LMK when you are satisfied with coverage and I'll merge. We can take care of any additional needs as follow-on work.

**Comment by beejeebus** [comment]
I'm happy enough with this as it is, follow ups for anything we missed works for me!

**Comment by greg-1-anderson** [comment]
Thanks for all of your work on this, it should be useful.

**Comment by weitzman** [comment]
If anyone has a chance to make a PR for Drush that would be great.

**Pull request merged**
